cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,488 @@
+#!/usr/bin/env python
+#-*- coding:utf-8 -*-
+##
+## exact.py
+##
+"""
+    Interface to Exact
+
+    Exact solves decision and optimization problems formulated as integer linear programs. Under the hood, it converts integer variables to binary (0-1) variables and applies highly efficient propagation routines and strong cutting-planes / pseudo-Boolean conflict analysis.
+
+    https://gitlab.com/JoD/exact
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_exact
+"""
+import sys  # for stdout checking
+
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..expressions.core import *
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl, intvar
+from ..transformations.comparison import only_numexpr_equality
+from ..transformations.flatten_model import flatten_constraint, flatten_objective, get_or_make_var
+from ..transformations.get_variables import get_variables
+from ..transformations.linearize import linearize_constraint, only_positive_bv, only_const_rhs, only_var_lhs
+from ..transformations.reification import only_bv_implies, reify_rewrite
+import numpy as np
+
+class CPM_exact(SolverInterface):
+    """
+    Interface to the Python interface of Exact
+
+    Requires that the 'exact' python package is installed:
+    $ pip install exact
+    See https://pypi.org/project/exact for more information.
+
+    Creates the following attributes (see parent constructor for more):
+    xct_solver: the Exact instance used in solve()
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import exact
+            return True
+        except ImportError as e:
+            return False
+
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Requires a CPMpy model as input, and will create the corresponding
+        Exact solver object xct_solver.
+
+        xct_solver can be modified externally before
+        calling solve(), a prime way to use more advanced solver features
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: None
+        """
+        if not self.supported():
+            raise Exception("Install the python 'exact' package to use this solver interface")
+
+        from exact import Exact as xct
+
+        assert subsolver is None, "Exact does not allow subsolvers."
+
+        # initialise the native solver object
+        self.xct_solver = xct()
+
+        # for solving with assumption variables,
+        self.assumption_dict = None
+
+        # objective can only be set once, so keep track of this
+        self.has_objective = False
```

Comment:
there might be a function has_objective() since a recent commit, maybe use has_objective_ as var in that case

Reply:
I can't seem to find (variations on) `has_objective()`. Renamed to `objective_is_set` to avoid name clashes.